doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,297 @@
+
+### Language
+
+We would like to keep the code for this project in Ruby as much as
+possible. Search API is written in Ruby and it is the primary language
+used at GOV.UK. Necessarily some machine learning code will need to be
+written in Python because TensorFlow is a Python library.
+
+### Dockerised model
+
+We will use TensorFlow Serving to make a re-ranker model available
+to Search API on search machines.
+
+We will run a docker container (containing our pre-trained model) on
+search machines to which Search API can send requests.
+
+Should the docker container fail, Search API should be resilient to this,
+and fall back to the default ranking algorithm, with results not
+re-ranked by the model.
+
+### Training
+
+**Training the model** should happen on search machines on a schedule.
+An instance of Search API will train a model using the latest relevancy
+judgements stored in S3, and will then upload a new version of the model to S3.
+On a schedule all search machines will pull the latest version of the model
+and push the new version to the locally running TF Serving container.
+
+**Training data** is made up of relevancy judgements and document features.
+The judgements may be sourced from user behaviour (implicit signals
+of relevancy) perhaps with a Click Model; so far we have used explicit
+manual judgements.
+
+Acquiring relevancy judgements is a work in progress. Our next project
+will be to create a process by which we can acquire relevancy judgements
+on a rolling basis automatically.
+
+### Re-ranking requests
+
+**Using the model** to re-rank results will involve a user searching,
+Search API will perform the query for the top-k results (say 100),
+next Search API will make a request for new scores from the re-ranker
+for the results, finally Search API will provide the re-ranked top-n
+results to the user.
+
+We may see greater search query latency when using the model. Search
+API will need to pull more data out of elasticsearch and do more
+computation, in addition to making a request to the re-ranker.
+This is something we will monitor, as we don't want to provide a much
+slower search experience at a cost of improving relevance.
+
+### Monitoring
+
+We are monitoring nDCG and click-through rates in a search relevancy
+dashboard in grafana.
+
+We will set up more monitoring and alerts as we get further into the project.
+
+### Testing the model
+
+We hope to re-use our existing offline nDCG tests and online AB testing
+process for Learning To Rank.
+
+AB testing models should be possible as TF Serving supports
+versioning models and can serve multiple models from one container.
+We plan to use AB testing to validate that the model provides better
+results to users (measured by click-through rate).
+
+## What success will look like
+
+Success will look like:
+
+- Click-through rate on the top 3 and top 5 results will go up
+- nDCG will go up
+- Search latency will not increase much
```

Comment:
This is probably still to be defined, but it'd be good to specify how much `much` is here.